### ISSUE TITLE
Track banned peers via IP/PORT

### ISSUE THREAD

**Issue by AgeManning** [sigp/lighthouse#1283]
## Description

Peers are identified via their PeerId. Their score is tagged to this `PeerId`. 

The tracking logic should be updated to keep track of PeerId and IP/PORT. We then blacklist not just a PeerId but also an IP/PORT to prevent attackers from spamming new `PeerId` and performing the same attack.

**Comment by ethDreamer** [comment]
This looks interesting. I'll check it out this weekend :)

**Comment by ethDreamer** [comment]
After looking into it, I'm not sure if this problem is straight-forward or if it involves more work to be done optimally.

Here's a simple way to do this. Let me know if this is on the right track with what you were thinking:
```
diff --git a/beacon_node/eth2_libp2p/src/peer_manager/peerdb.rs b/beacon_node/eth2_libp2p/src/peer_manager/peerdb.rs
index 063d36db..1e18059f 100644
--- a/beacon_node/eth2_libp2p/src/peer_manager/peerdb.rs
+++ b/beacon_node/eth2_libp2p/src/peer_manager/peerdb.rs
@@ -1,18 +1,20 @@
 use super::peer_info::{PeerConnectionStatus, PeerInfo};
 use super::peer_sync_status::PeerSyncStatus;
 use crate::rpc::methods::MetaData;
 use crate::PeerId;
 use slog::{crit, debug, trace, warn};
 use std::collections::{hash_map::Entry, HashMap};
 use std::time::Instant;
 use types::{EthSpec, SubnetId};
+use crate::peer_manager::Multiaddr;
+use std::collections::HashSet;
 
 /// A peer's reputation (perceived potential usefulness)
 pub type Rep = u8;

 
@@ -23,20 +25,22 @@ const MAX_DC_PEERS: usize = 30;
 pub const DEFAULT_REPUTATION: Rep = 50;
 
 /// Storage of known peers, their reputation and information
 pub struct PeerDB<TSpec: EthSpec> {
     /// The collection of known connected peers, their status and reputation
     peers: HashMap<PeerId, PeerInfo<TSpec>>,
     /// Tracking of number of disconnected nodes
     n_dc: usize,
     /// PeerDB's logger
     log: slog::Logger,
+    /// Banned Addresses
+    banned_addrs: HashSet<Multiaddr>,
 }

@@ -52,20 +56,21 @@ impl RepChange {
         }
     }
 }
 
 impl<TSpec: EthSpec> PeerDB<TSpec> {
     pub fn new(log: &slog::Logger) -> Self {
         Self {
             log: log.clone(),
             n_dc: 0,
             peers: HashMap::new(),
+            banned_addrs: HashSet::new(),
         }
     }
 
@@ -101,22 +106,34 @@ impl<TSpec: EthSpec> PeerDB<TSpec> {
     pub fn peer_synced(&self, peer_id: &PeerId) -> bool {
         match self.peers.get(peer_id).map(|info| &info.sync_status) {
             Some(PeerSyncStatus::Synced { .. }) => true,
             Some(_) => false,
             None => false,
         }
     }
 
     /// Returns true if the Peer is banned.
     pub fn peer_banned(&self, peer_id: &PeerId) -> bool {
-        match self.peers.get(peer_id).map(|info| &info.connection_status) {
-            Some(status) => status.is_banned(),
+        match self.peers.get(peer_id) {
+            Some(info) => {
+                if info.connection_status.is_banned() {
+                    true
+                }
+                else {
+                    for address in info.listening_addresses.iter() {
+                        if self.banned_addrs.contains(address) {
+                            return true;
+                        }
+                    }
+                    false
+                }
+            },
+            // See note below about this case
             None => false,
         }
     }
 
@@ -335,20 +352,25 @@ impl<TSpec: EthSpec> PeerDB<TSpec> {
     pub fn ban(&mut self, peer_id: &PeerId) {
         let log_ref = &self.log;
         let info = self.peers.entry(peer_id.clone()).or_insert_with(|| {
             warn!(log_ref, "Banning unknown peer";
                     "peer_id" => peer_id.to_string());
             PeerInfo::default()
         });
         if info.connection_status.is_disconnected() {
             self.n_dc = self.n_dc.saturating_sub(1);
         }
+        // Ban every known listening address
+        for address in info.listening_addresses.iter() {
+            self.banned_addrs.insert(address.clone());
+        }
+
         info.connection_status.ban();
     }
 
```
The only place where `peer_banned()` is checked is in [peers_discovered()](https://github.com/sigp/lighthouse/blob/master/beacon_node/eth2_libp2p/src/peer_manager/mod.rs#L425) so I believe this would work once peers are actually banned. Currently, no peers ever get banned because calls to [PeerManager.update_reputations()](https://github.com/sigp/lighthouse/blob/master/beacon_node/eth2_libp2p/src/peer_manager/mod.rs#L508) aren't made yet and the code for this function is commented out.

I did notice [this comment](https://github.com/sigp/lighthouse/blob/master/beacon_node/eth2_libp2p/src/peer_manager/mod.rs#L421) on `peers_discovered()` which suggests this might be refactored to use addresses rather than peer_id's in which case there might be more work to do this properly...

Along that line of thinking, I'm not sure if it's possible (in the future) for `peer_banned()` to be called before the peer_id is in the peerDB. If it were possible, the last match case (None) should run the same check by grabbing the addresses using a similar logic as [addresses_of_peer()](https://github.com/sigp/lighthouse/blob/master/beacon_node/eth2_libp2p/src/peer_manager/mod.rs#L384) perhaps... though without access to the `PeerManager` this isn't really possible.. just a thought I had

anyway.. would appreciate some feedback on this :)

**Comment by AgeManning** [comment]
Hey, great work! 

Apologies for the little detail in this issue. I have a quite a few issues similar to this one, where there is little detail. I'm kind of using these as notes to myself as I go, to remember to go back and add in the feature. 

I was out this weekend, so sorry for not responding sooner. 

As you've discovered there is a bit of code here that isn't currently in use. In particular the banning and using the banning logic isn't connected and has no real effect on current lighthouse. #1284 will introduce this logic and integrate the banning and peer scoring logic into lighthouse. So this issue is currently kind of blocked with #1284. 

In terms of what you've done, I think I agree with you. We would need to ban IP/Multiaddrs separately from peers in the peer manager. There is some added complexity however, as I've currently designed the scoring system to have a decay rate, where banned peer's scores will decay over time and become usable again. 

I think we'd need to think about how banning IP's would work also. Some clients are using relays, in which many peers use a relay peer with a single IP. If one peer misbehaves, should we ban the IP for a specific amount of time?

I think in the first iteration of this, we have some hard coded constant, `NUMBER_OF_BANNED_PEERS_PER_IP` which when we exceed this, i.e we ban a peer on the same IP more than this number of times, then we ban the IP for a fixed amount of time. 

Let me finish #1284 (which should be done this week, if not too many other things come up) then we should be able to add this guy in also. :) 
I imagine it will be clearer where everything fits into place then. I'll be able to link some code with where it's being used also.

**Comment by divagant-martian** [comment]
Is this still blocked?

**Comment by blacktemplar** [comment]
I don't think so, I am currently working on it and will submit a pull request today :).

**Comment by AgeManning** [comment]
Closed in #1543
